device.c

```diff
@@ -142,13 +161,12 @@ struct device *device_open(const char *board,
 
 static void device_impl_power(struct device *device, bool on)
 {
-	device->power(device, on);
+	device_control(device, power, on);
 }
 
 static void device_key(struct device *device, int key, bool asserted)
 {
-	if (device->key)
-		device->key(device, key, asserted);
+	device_control_opt(device, key, key, asserted);
```

Comment:
To me this is a lot less readable, without the macro it's clear what exactly this code does, but now I would have to go and find the definition of `device_control_opt` and understand it.

What about:

```c
if (device_has_control(key))
    device->key(device, key, asserted);
```

Reply:
It becomes:
```c
if (device_has_control(key))
    device->control_ops->key(device, key, asserted);
```

The latter part was what @superna9999 disliked, too many indirections.

Reply:
@calebccff reworking this to be more obvious.